### PR TITLE
Federated averaging with local and global optimizations

### PR DESCRIPTION
**Pull request by arthurPignet**
This PR follows the draft 235, which was too far behind master. This is a rebased version.  

# New mpl methods:

FedGDO stands for Federated Gradient Double Optimization.

This method is inspired from Federated gradient, but with modification on the local computation of the gradient.
In this version we use a local optimizer (partner-specific) to do several minimization steps of the local-loss
during a minibatch. We use the sum of these weighs-updates as the gradient which is sent to the global optimizer.
The global optimizer aggregates these gradients-like which have been sent by the partners,
and performs a optimization step with this aggregated gradient.
Here three variations of this mpl method are tested.
 - FedGDO with fresh local optimizer at each minibatch
 - FedGDO with persistent local optimizer

The reset of the optimizers can be set via a parameter of the mpl methods. In the same way, a global optimizer different from the local ones can be passed via the mpl arguments. 

These methods are tested on this notebook -> https://colab.research.google.com/drive/1CcQpWRpLGldj3iNR7v7Hv2brdwBP3z7D?usp=sharing

Please note that as I am currently working with the notebook, it can change, and be not fully readable 
Notebook access is currently limited to substra.org, but don't hesitate to come to me for access.

### CODE REVIEW

**Review by bowni — APPROVED** [review]
OK, well done @arthurPignet this looks very promising (although not always easy to picture and understand line by line) 💪 
@RomainGoussault @jeromechambost @celinejacques @SaboniAmine @HeytemBou @JustineBoulant @kat-leen who would like to do another review please?

**Comment by bowni** [comment]
@RomainGoussault proposes to review it!

File: mplc/multi_partner_learning/fast_mpl.py [hunk 1]
```diff
@@ -575,3 +575,95 @@ def fit_epoch(model, train_dataset, partners_grads, smodel_list, global_grad, ag
                 break
 
         self.log_end_training()
+
+
+class FastFedGDO(FastFedAvg):
+    """
+     This method is inspired from Federated gradient, but with modification on the local computation of the gradient.
```
Comment (RomainGoussault) [inline mplc/multi_partner_learning/fast_mpl.py:582]: I would add that FedGDO stand for Federated averaging with double optimizers.

File: mplc/multi_partner_learning/fast_mpl.py [hunk 2]
```diff
@@ -575,3 +575,95 @@ def fit_epoch(model, train_dataset, partners_grads, smodel_list, global_grad, ag
...
+
+    def __init__(self, scenario, reset_local_optims=False, global_optimiser=None, **kwargs):
+        self.global_optimiser = global_optimiser
+        self.reset_local_optims = reset_local_optims
+        super(FastFedGDO, self).__init__(scenario, **kwargs)
+
+    def init_specific_tf_variable(self):
+        # generate tf Variables in which we will store the model weights
+        self.model_stateholder = [tf.Variable(initial_value=w.read_value()) for w in self.model.trainable_weights]
+        self.partners_grads = [[tf.Variable(initial_value=w.read_value()) for w in self.model.trainable_weights]
```
Comment (RomainGoussault) [inline mplc/multi_partner_learning/fast_mpl.py:598]: this is gradients or weights?

File: mplc/multi_partner_learning/fast_mpl.py [hunk 3]
```diff
@@ -575,3 +575,95 @@ def fit_epoch(model, train_dataset, partners_grads, smodel_list, global_grad, ag
...
+                               for _ in self.partners_list]
+        self.global_grad = [tf.Variable(initial_value=w.read_value()) for w in self.model.trainable_weights]
+        self.partners_optimizers = [self.model.optimizer.from_config(self.model.optimizer.get_config()) for _ in
+                                    self.partners_list]
+        if self.global_optimiser:
```
Comment (RomainGoussault) [inline mplc/multi_partner_learning/fast_mpl.py:603]: in which case we won't have a global optimizer?

File: mplc/multi_partner_learning/fast_mpl.py [hunk 4]
```diff
@@ -575,3 +575,95 @@ def fit_epoch(model, train_dataset, partners_grads, smodel_list, global_grad, ag
...
+            # at the end of the minibatch, aggregate all the local grads
+            for i, grads_per_layer in enumerate(zip(*partners_grads)):
+                global_grad[i].assign(tf.tensordot(grads_per_layer, aggregation_weights, [0, 0]))
+
+            # perform one optimization update using the aggregated gradient
+            model.optimizer.apply_gradients(
+                zip(global_grad, model.trainable_weights))
+
+            # Execution
+
```
Comment (RomainGoussault) [inline mplc/multi_partner_learning/fast_mpl.py:646]: blanck line to remove?

**Review by RomainGoussault — CHANGES_REQUESTED** [review]

**Comment by bowni** [comment]
@arthurPignet let us know when you can tackle @RomainGoussault 's questions and comments! 😃

File: mplc/multi_partner_learning/fast_mpl.py (hunk 2 above)
Comment (arthurPignet) [reply mplc/multi_partner_learning/fast_mpl.py:598]: The first one is for the weights, to store initial weight of the global network at the beginning of the global batch while one partner is optimizing the network within its local batch
The second one is for gradients 

But I use the weights for both because it is convenient as the size of gradients is the same as the size of weights.

File: mplc/multi_partner_learning/fast_mpl.py (hunk 1 above)
Comment (arthurPignet) [reply mplc/multi_partner_learning/fast_mpl.py:582]: 👍

File: mplc/multi_partner_learning/fast_mpl.py (hunk 3 above)
Comment (arthurPignet) [reply mplc/multi_partner_learning/fast_mpl.py:603]: When the global optimizer is not specified in the kwargs. In that case the global optimizer used is the default one, same as the local optimizer, and we don't need to re-compile the model

File: mplc/multi_partner_learning/fast_mpl.py (hunk 4 above)
Comment (arthurPignet) [reply mplc/multi_partner_learning/fast_mpl.py:646]: 👍

File: mplc/multi_partner_learning/fast_mpl.py [hunk 5]
```diff
@@ -575,3 +575,95 @@ def fit_epoch(model, train_dataset, partners_grads, smodel_list, global_grad, ag
...
+                    with tf.GradientTape() as tape:
+                        y_pred = model(x)
+                        loss = model.compiled_loss(y, y_pred)
+                    model.compiled_metrics.update_state(y, y_pred)  # log the loss and accuracy
+                    partners_optimizers[p_id].minimize(loss, model.trainable_weights,
+                                                       tape=tape)  # perform local optimizations
+                # get the gradient as theta_before_minibatch - theta_after_minibatch
+                for grad_per_layer, w_old, w_new in zip(partners_grads[p_id], model_stateholder,
+                                                        model.trainable_weights):
+                    grad_per_layer.assign((w_old - w_new))
```
Comment (HeytemBou) [inline mplc/multi_partner_learning/fast_mpl.py:632]: @arthurPignet Here you're using weight differences between two consecutive minibatch updates as gradients, right? but how does that work on a theoretical level?

Comment (arthurPignet) [reply mplc/multi_partner_learning/fast_mpl.py:632]: Yes indeed. 
When you are performing a classic SGD update, the difference between the weights is the stocastic estimation of the gradient * lr. 
But here we are using more sophisticated optimization algorithms, with momentums and adaptative lr. The gradient is not easy to reach. So I fall back to an object which is not exactly the gradient, but the cumulative descent direction (which is the object of our interest actually). I compute the descent direction to be sent to the server as the vec. sum of the descents applied during the local batches. That corresponds to the difference between the weights at the beginning of the local batches and the ones at the ends of the local batches, just before the aggregation/global optimization which will conclude the global batch.  

Did I answer your question ? Maybe I misunderstood what you were saying.

Comment (HeytemBou) [reply mplc/multi_partner_learning/fast_mpl.py:632]: Yes thanks, it's easy to miss that the difference between the weights equates to gradient*lr, unless you have the SGD update rule in front of your eyes :). parhaps it would be much clearer for future readers to  to change the name of the parameter "global_grad" to something like cumulative_descent maybe, just like you explained. It's a bit misleading at first when reading the name 'grad' being used.

Comment (arthurPignet) [reply mplc/multi_partner_learning/fast_mpl.py:632]: Ok that is a nice idea. 
But I will use this cumulative_descent like a gradient just after for the aggregation and the global optimization. Don't you think that using a object which is not named grad where it is usually a gradient in every other algorithm, to play the role of the gradient could be even more misleading ?
Maybe rather than a reformating the variable I could add a comment saying that this thing is not really a gradient but a cumulative descent ?

Comment (HeytemBou) [reply mplc/multi_partner_learning/fast_mpl.py:632]: yes, a comment would be great.

Comment (bowni) [reply mplc/multi_partner_learning/fast_mpl.py:632]: Or name it `pseudo_gradient` or something of that sort

**Comment by bowni** [comment]
@HeytemBou beyond your above comment and the associated thread, was your review conclusive? Are you confortable with validating the PR or not?

**Comment by HeytemBou** [comment]
> @HeytemBou beyond your above comment and the associated thread, was your review conclusive? Are you confortable with validating the PR or not?

Yes

**Review by HeytemBou — APPROVED** [review]

**Comment by arthurPignet** [comment]
> @arthurPignet let us know when you can tackle @RomainGoussault 's questions and comments! 😃

I took into account @RomainGoussault's comments, it's ready for a final (well I hope it will be) review !

File: mplc/multi_partner_learning/fast_mpl.py (hunk 3 above)
Comment (RomainGoussault) [reply mplc/multi_partner_learning/fast_mpl.py:604]: Maybe we can add a small comment to say what you just wrote?

**Review by RomainGoussault — APPROVED** [review]
LGTM8
